### PR TITLE
OpenPGP support: Compose Logic

### PR DESCRIPTION
**Pull request by Valodim**
This is my working branch for OpenPGP support in the compose view. It is based off @alexandre75's work, and so far I have added key-awareness to the recipient picker, redesigning those parts of the UI into a TokenCompleteTextView workflow and applying an MVP pattern to relevant parts of MessageCompose in the process.

Note that this branch requires a development branch of OpenKeychain to communicate with: https://github.com/open-keychain/open-keychain/pull/1615

As mentioned this is a WIP PR. I'll try to keep it up to date about my progress, and do feel free to give feedback of course.

### CODE REVIEW

**Comment by Valodim** [comment]
A lot happened in these last three commits:

first:
* the SendMessageTask no longer builds the message, it performs only the things which are actually related to sending the message (delete draft, send, mark contacts as contacted)
* the MessageBuilder build() call is now buildAsync(), and delivers the result back to the main thread in a callback (though this is not actually done asynchronously in this commit)
* the pgp/mime construction logic now lives in the PgpMessageBuilder
* pending intents of the OpenPgpApi are handled as intended
* the MessageBuilder class is *not* able to do onSaveInstanceState (see below)

second:
* during encryption and signing, stream data to and from the openpgpapi instead of reading it all into ram
* this required changes to the OpenPgpApi, which is why I re-included that in the plugin/ folder in an earlier commit

third:
* work on async properties of MessageBuilder. the callback can now be detached intermittently, postponing result delivery until it is reattached
* at this point, only the SimpleMessageBuilder actually runs asynchronously, I wanted to get the lifecycle stuff right before I add this to PgpMessageBuilder
* for now, the activity keeps the MessageBuilder as a nonConfigurationInstance state during configuration changes. this is a deprecated method, but since we aren't using fragments or anything it seems like the way to go here

**Comment by Valodim** [comment]
rebased this on top of #962

**Comment by dschuermann** [comment]
To prevent inclusion of the OpenPGP api lib again, I would propose we release a new version to JCenter before this pull request is ready to be merged, i.e., no more changes to the lib are required.

**Comment by Valodim** [comment]
Hmm I think I would prefer if we kept it, as there will likely be more changes down the line. I'll have to keep it in tree locally either way for development and making a release per intermediate state seems like a hassle for little benefit.

**Comment by cketti** [comment]
To not slow down development I think it's okay to include the library for now. But let's add an issue to the PGP/MIME milestone to remove it from the source tree again. We don't want two versions of the library in the long run.

**Comment by Valodim** [comment]
rebased on top of current master and worked on some more aspects:
* SignSafeOutputStream now only "escapes" a From at the beginning of a line, since end-of-line whitespace escaping is already done by QuotedPrintableOutputStream
* removed content-type magic from MimeMessage, just statically set the header in the message builder for now
* added "private" mode, for sending only if keys are available for all recipients
* all crypto info is accumulated in an immutable ComposeCryptoStatus, which is used as info in the view and also passed to the message builder
* actually run pgpmessagebuilder operations in worker thread, unify return methods
* use opaque "key reference" field from provider to hand encrypt-to info to openkeychain, rather than user ids. this way, the displayed data is closer to the one actually used by openkeychain

there are a couple of things still missing, most are fairly minor:
* there is no indicator at the moment while the message is being processed in the background, didn't look at how to best do that yet
* drafts should be stored encrypted. this is not in yet, since it requires decryption of messages when a draft is edited I will come back to this once I worked on message decryption
* icons need improvement, but I'm not sure how I really want these.. I'll just leave that open for now

**Comment by Valodim** [comment]
other than the "better icons" part, I think this PR should be good at this point. code reviews welcome.

File: k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java [hunk 1]
```diff
@@ -318,13 +307,8 @@ public void onCryptoModeChanged(CryptoMode cryptoMode) {
...
     private QuoteStyle mQuoteStyle;
 
-    private boolean mDraftNeedsSaving = false;
-    private boolean mPreventDraftSaving = false;
-
-    /**
-     * If this is {@code true} we don't save the message as a draft in {@link #onPause()}.
-     */
-    private boolean mIgnoreOnPause = false;
+    private boolean draftNeedsSaving = false;
```
Comment (dschuermann) [inline k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java]: just for consistency: these are not preceded by 'm', the other private variables are. is there a reason?

Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java]: style guideline says not to use m, so that's what I do for new variables (unless I forget)

File: k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java [hunk 2]
```diff
@@ -3253,12 +2997,56 @@ private boolean isCryptoProviderEnabled() {
...
 
-    private boolean shouldSign() {
-        return isCryptoProviderEnabled() && mCryptoSignatureCheckbox.isChecked();
+    @Override
+    public void onMessageBuildReturnPendingIntent(PendingIntent pendingIntent, int requestCode) {
+        requestCode |= REQUEST_MASK_MESSAGE_BUILDER;
+        try {
+            startIntentSenderForResult(pendingIntent.getIntentSender(), requestCode, null, 0, 0, 0);
+        } catch (SendIntentException e) {
+            e.printStackTrace();
```
Comment (dschuermann) [inline k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java]: Log.e

File: k9mail/src/main/java/com/fsck/k9/ui/crypto/MessageCryptoHelper.java [hunk 3]
```diff
@@ -36,6 +36,7 @@
 import com.fsck.k9.mailstore.OpenPgpResultAnnotation;
 import com.fsck.k9.mailstore.OpenPgpResultAnnotation.CryptoError;
 import org.openintents.openpgp.IOpenPgpService;
```
Comment (dschuermann) [inline k9mail/src/main/java/com/fsck/k9/ui/crypto/MessageCryptoHelper.java]: this import still needed?

Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/ui/crypto/MessageCryptoHelper.java]: maybe not - it's not part of the PR either way :)

**Comment by dschuermann** [comment]
I read the code from top to bottom, looks good. Will do some testing tomorrow

File: k9mail-library/src/androidTest/java/com/fsck/k9/mail/internet/MimeHeaderTest.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+
+import static junit.framework.Assert.assertEquals;
+import static junit.framework.Assert.assertTrue;
+
+/**
+ * Created by alexandre on 11/5/15.
+ */
```
Comment (dschuermann) [inline k9mail-library/src/androidTest/java/com/fsck/k9/mail/internet/MimeHeaderTest.java]: unneeded comment

**Comment by Valodim** [comment]
(rebased on current master after the latest merges)

**Comment by cketti** [comment]
"Always Sign, Encrypt if possible" doesn't work for me. It encrypts the message to the address I'm sending from, but not the recipient. Also, the message is not signed.
In the recipient view the address I'm sending to is displayed with a red lock even though I have a trusted key for that address.

**Comment by Valodim** [comment]
The first part is a thing in openkeychain which I never pushed because I only applied a workaround and uh, then forgot. I'll get back to that.

About the mail not being signed, I'm a little stumped. Are you sure the secret key is available? Should be relatively simple to figure out what's wrong by stepping through `startOrContinueBuildMessage` and looking at the ComposeCryptoStatus object.

**Comment by cketti** [comment]
Actually, the message seems to be signed correctly. I only looked at the sent message in K-9 Mail, which doesn't support validating a separately encrypted part yet :(

**Comment by Valodim** [comment]
The part about encryption is sort of an open point in OpenKeychain. I pushed my incomplete solution to the external-provider branch to have a working version online, but that behavior is still subject to changes. I tried to decouple this behavior as much as possible from K-9 though.
